### PR TITLE
Variable size in EC2 API request and ifdefs for linux build

### PR DESCRIPTION
**Pull request by ZackaryWelch**

### CODE REVIEW

**Comment by ZackaryWelch** [comment]
https://github.com/cperciva/bsdec2-image-upload/issues/9

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 *.d
 *.o
 bsdec2-image-upload
+.depend
```
Comment (cperciva) [inline .gitignore:4]: This looks like it belonged in the previous commit.

File: lib/util/sslreq.c [hunk 2]
```diff
@@ -1,6 +1,10 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
+#ifdef __linux__
```
Comment (cperciva) [inline lib/util/sslreq.c]: For portability, make this (and the call which uses it later) `#ifndef SO_NOSIGPIPE`.

File: lib/util/sslreq.c [hunk 3]
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
 	if (r == NULL)
 		return "Could not connect";
 
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /* Linux does not support SO_NOSIGPIPE
```
Comment (cperciva) [inline lib/util/sslreq.c]:
```
/*
 * Multi-line comments should
 * be formatted like this.
 */
```

Comment (cperciva) [reply lib/util/sslreq.c]: Also, sentences end with `.`.

File: lib/util/sslreq.c [hunk 4]
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
...
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally */
+	#ifdef __linux__
```
Comment (cperciva) [inline lib/util/sslreq.c]: `#ifndef SO_NOSIGPIPE`, and also preprocessor directives should be in column 1.

File: lib/util/sslreq.c [hunk 5]
```diff
@@ -153,9 +157,15 @@ sslreq2(const char * host, const char * port, const char * certfile,
...
+		signal(SIGPIPE, SIG_IGN);
```
Comment (cperciva) [inline lib/util/sslreq.c]: Error handling?

File: main.c [hunk 6]
```diff
@@ -21,8 +21,13 @@
 #include "warnp.h"
 
 #ifndef CERTFILE
+/* Different cert locations between Linux and BSD */
```
Comment (cperciva) [inline main.c]: This should be handled by defining CERTFILE in the `make` invocation.

File: main.c [hunk 7]
```diff
@@ -1256,12 +1261,22 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
...
 
-	/* Map the file. */
-	if ((p = mmap(NULL, len, PROT_READ, MAP_PRIVATE | MAP_NOCORE,
-	    fileno(f), 0)) == MAP_FAILED) {
-		warnp("Could not map disk image");
-		goto err1;
-	}
+
+    /* Linux does not support MAP_NOCORE */
+    #ifdef __linux__
```
Comment (cperciva) [inline main.c]: I would change this to
```
#ifndef MAP_NOCORE
#define MAP_NOCORE 0
#endif
```

File: main.c [hunk 8]
```diff
@@ -1334,7 +1349,7 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
 static char *
 registerimage(const char * region, const char * snapshot, const char * name,
     const char * desc, const char * arch, int sriov, int ena,
-    const char * key_id, const char * key_secret)
+    const char * key_id, const char * key_secret, const char * diskimage)
```
Comment (cperciva) [inline main.c]: We're getting the file size in uploadvolume and uploadsnap.  Return the size from those functions and pass it into this one rather than statting the file a second time.

Comment (ZackaryWelch) [reply main.c]: How should this be done?

Comment (cperciva) [reply main.c]: Add a "imgsz" parameter to each of the functions?

File: Makefile [hunk 9]
```diff
@@ -2,10 +2,18 @@ PROG=	bsdec2-image-upload
...
 MAN	=
 WARNS	?=	3
-CFLAGS	+=	-Wno-error=\#warnings
 BINDIR	?=	/usr/local/bin
 LDADD	+=	-lcrypto -lssl
 
+UNAME_S := $(shell uname -s)
+.if (${UNAME_S},Linux)
+CERTFILE = /etc/ssl/certs/ca-bundle.crt
+CFLAGS += --no-warnings
```
Comment (cperciva) [inline Makefile]: Umm... no.

File: Makefile [hunk 10]
```diff
@@ -40,5 +48,6 @@ IDIRS	+=	-I lib/util
 
 CFLAGS	+=	-g
 CFLAGS	+=	${IDIRS}
+CFLAGS  += -DCERTFILE
```
Comment (cperciva) [inline Makefile]: What do you think this does, exactly?

File: lib/util/sslreq.c (hunk 2 above)
Comment (cperciva) [reply lib/util/sslreq.c]: Still need to fix this.

File: lib/util/sslreq.c [hunk 11]
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
...
 
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /*
+     * Linux does not support SO_NOSIGPIPE.
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally.
+     */
+#ifndef SO_NOSIGPIPE
+		if (signal(SIGPIPE, SIG_IGN) == SIG_ERR)
```
Comment (cperciva) [inline lib/util/sslreq.c]: Superfluous indentation.

File: lib/util/sslreq.c [hunk 12]
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
 	if (r == NULL)
 		return "Could not connect";
 
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /*
+     * Linux does not support SO_NOSIGPIPE.
+     * To accomplish the same effect, signal is used, which disables SIGPIPE globally.
```
Comment (cperciva) [inline lib/util/sslreq.c]: Please rewrap.  There's no need for each sentence to be on its own line.

File: main.c [hunk 13]
```diff
@@ -1256,12 +1258,12 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
 	}
 	len = sb.st_size;
 
-	/* Map the file. */
-	if ((p = mmap(NULL, len, PROT_READ, MAP_PRIVATE | MAP_NOCORE,
-	    fileno(f), 0)) == MAP_FAILED) {
-		warnp("Could not map disk image");
-		goto err1;
-	}
+    /* Map the file. */
```
Comment (cperciva) [inline main.c]: Why mangle the indentation?

File: Makefile (hunk 10 above)
Comment (ZackaryWelch) [reply Makefile:51]: https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html

File: Makefile (hunk 9 above)
Comment (ZackaryWelch) [reply Makefile]: How would you suggest defining CERTFILE in the makefile?

File: Makefile (hunk 10 above)
Comment (ZackaryWelch) [reply Makefile]: Oh lol, I'm missing the assignment

File: Makefile [hunk 14]
```diff
@@ -2,10 +2,18 @@ PROG=	bsdec2-image-upload
...
 MAN	=
 WARNS	?=	3
-CFLAGS	+=	-Wno-error=\#warnings
 BINDIR	?=	/usr/local/bin
 LDADD	+=	-lcrypto -lssl
 
+UNAME_S	:=	$(shell uname -s)
+.if (${UNAME_S},Linux)
+CERTFILE	=	\"/etc/ssl/certs/ca-bundle.crt\"
+CFLAGS	+=	--no-warnings
```
Comment (cperciva) [inline Makefile:11]: Still not sure why you're disabling all compiler warnings on Linux.

File: lib/util/sslreq.c [hunk 15]
```diff
@@ -153,9 +157,18 @@ sslreq2(const char * host, const char * port, const char * certfile,
 	if (r == NULL)
 		return "Could not connect";
 
-	/* Disable SIGPIPE on this socket. */
-	if (setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, &on, sizeof(on)))
-		return "Could not disable SIGPIPE";
+    /*
```
Comment (cperciva) [inline lib/util/sslreq.c:160]: Please use tabs for indentation, like the rest of the file does.

File: main.c [hunk 16]
```diff
@@ -1255,10 +1259,11 @@ uploadsnap(const char * fname, const char * imgfmt, const char * region,
 		goto err1;
 	}
 	len = sb.st_size;
+	*imgsz = len;
 
 	/* Map the file. */
 	if ((p = mmap(NULL, len, PROT_READ, MAP_PRIVATE | MAP_NOCORE,
-	    fileno(f), 0)) == MAP_FAILED) {
+		fileno(f), 0)) == MAP_FAILED) {
```
Comment (cperciva) [inline main.c:1266]: Why did you change the indentation here?

File: main.c [hunk 17]
```diff
@@ -1350,6 +1355,10 @@ registerimage(const char * region, const char * snapshot, const char * name,
 		goto err1;
 	if ((archenc = rfc3986_encode(arch)) == NULL)
 		goto err2;
+    
+    /* Get image size in GB and set to 10 if less than 10. */
+    size_t gb_size = (*imgsz)/(1024 * 1024 * 1024);
```
Comment (cperciva) [inline main.c:1360]: Put variable declarations at the top of the function.

Also, the indentation is wrong.

Also, there should be whitespace on either side of the `/` character.

Also, `if(gb_size < 10) gb_size = 10;` should be
```
if (gb_size < 10)
        gb_size = 10;
````

File: main.c [hunk 18]
```diff
@@ -1364,7 +1373,7 @@ registerimage(const char * region, const char * snapshot, const char * name,
 	    "BlockDeviceMapping.1.DeviceName=%%2Fdev%%2Fsda1&"
 	    "BlockDeviceMapping.1.Ebs.SnapshotId=%s&"
 	    "BlockDeviceMapping.1.Ebs.VolumeType=gp2&"
-	    "BlockDeviceMapping.1.Ebs.VolumeSize=10&"
+	    "BlockDeviceMapping.1.Ebs.VolumeSize=%d&"
```
Comment (cperciva) [inline main.c:1376]: `%d` is the wrong type for printing a `size_t` value.

File: main.c [hunk 19]
```diff
@@ -2102,14 +2111,16 @@ main(int argc, char * argv[])
 		nregions = 1;
 	}
 
+	size_t *imgsz = malloc(sizeof(size_t));
```
Comment (cperciva) [inline main.c:2114]: Umm, what?  Just have `size_t imgsz` and pass `&imgsz` to functions below.

File: Makefile (hunk 14 above)
Comment (ZackaryWelch) [reply Makefile:11]: What does -Wno-error=\\#warnings do? That was my best guess as to what it did.

File: main.c (hunk 16 above)
Comment (ZackaryWelch) [reply main.c:1266]: I have no idea how this happened

**Comment by ZackaryWelch** [comment]
Awesome! This all appears implemented now.
